local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,226 @@ fn test_slot_hash_expiry() {
+                .collect::<Vec<_>>(),
+        ),
+        node_vote_keys: Some(node_vote_keys),
+        slots_per_epoch,
+        stakers_slot_offset: slots_per_epoch,
+        skip_warmup_slots: true,
+        ..ClusterConfig::default()
+    };
+    let mut cluster = LocalCluster::new(&mut config, SocketAddrSpace::Unspecified);
+
+    let mut common_ancestor_slot = 8;
+
+    let a_ledger_path = cluster.ledger_path(&a_pubkey);
+    let b_ledger_path = cluster.ledger_path(&b_pubkey);
+
+    // Immediately kill B and C (we just needed it for the initial stake distribution)
+    info!("Killing B and C");
+    let mut b_info = cluster.exit_node(&b_pubkey);
+    cluster.exit_node(&c_pubkey);
+
+    // Let A run for a while until we get to the common ancestor
+    info!("Letting A run until common_ancestor_slot");
+    let now = Instant::now();
+    loop {
+        let last_vote = wait_for_last_vote_in_tower_to_land_in_ledger(&a_ledger_path, &a_pubkey);
+        if last_vote >= common_ancestor_slot {
+            common_ancestor_slot = last_vote;
+            info!("common_ancestor_slot: {}", common_ancestor_slot);
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+        sleep(Duration::from_millis(100));
+    }
+
+    // Copy A's ledger to B but truncate any blocks past the common ancestor.
+    {
+        info!("Copying A's ledger to B");
+        std::fs::remove_dir_all(&b_info.info.ledger_path).unwrap();
+        let mut opt = fs_extra::dir::CopyOptions::new();
+        opt.copy_inside = true;
+        fs_extra::dir::copy(&a_ledger_path, &b_ledger_path, &opt).unwrap();
+
+        // remove A's tower in B's new copied ledger
+        info!("Removing A's tower in B's ledger dir");
+        remove_tower(&b_ledger_path, &a_pubkey);
+
+        // Get rid of any slots past common_ancestor_slot
+        info!("Removing extra slots from B's blockstore");
+        let blockstore = open_blockstore(&b_ledger_path);
+        purge_slots_with_count(&blockstore, common_ancestor_slot + 1, 100);
+    }
+
+    info!("Run A on majority fork until it creates its own fork");
+    let mut last_vote_on_a;
+    // Keep A running for a while longer so A has its own fork.
+    let now = Instant::now();
+    loop {
+        last_vote_on_a = wait_for_last_vote_in_tower_to_land_in_ledger(&a_ledger_path, &a_pubkey);
+        if last_vote_on_a > common_ancestor_slot + 3 {
+            let blockstore = open_blockstore(&a_ledger_path);
+            info!(
+                "A majority fork: {:?}",
+                AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+            );
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+        sleep(Duration::from_millis(100));
+    }
```

Comment:
why not merge this entire section with the previous loop by checking `if last_vote >= common_ancestor_slot + 3` instead

Reply:
Changed.

Reply:
I mean, I simplified the previous loop, but the two loops are for different purpose, and there is a ledger copy in the middle. I think conceptually it's simpler to keep the two apart.